### PR TITLE
Feature: Track total supply

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Add feature flags to be able to specify whether total supply is tracked or not for resources

## Details
* Refactored PackageSetup
* Added feature flags to be able to enable/disable methods/substates/logic similar to rust `cfg(feature)`

### For dApp Developers
* One may manually disable tracking total supply

### For Internal Integrators
* Package Info substate has been updated

### CODE REVIEW

File: radix-engine-interface/src/blueprints/package/invocations.rs [hunk 1]
```diff
@@ -16,14 +16,14 @@ pub const PACKAGE_PUBLISH_WASM_IDENT: &str = "publish_wasm";
 #[derive(Debug, Clone, Eq, PartialEq, ScryptoSbor, ManifestSbor)]
 pub struct PackagePublishWasmInput {
     pub code: Vec<u8>,
-    pub definition: PackageDefinition,
+    pub setup: PackageSetup,
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/invocations.rs:19]: Prefer to have `PackageDefinition`.

File: radix-engine-interface/src/blueprints/package/invocations.rs [hunk 2]
```diff
@@ -94,8 +94,22 @@ pub struct PackageClaimRoyaltiesInput {}
...
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Default, ScryptoSbor, ManifestSbor)]
+pub struct BlueprintSetup {
+    pub schema: BlueprintSchema,
+
+    pub function_auth: BTreeMap<String, AccessRule>,
+    pub royalty_config: RoyaltyConfig,
+
+    pub template: BlueprintTemplate,
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/invocations.rs:108]: Suggest `method_auth_template: MethodAuthTemplate`

File: radix-engine-interface/src/blueprints/package/invocations.rs [hunk 3]
```diff
@@ -94,8 +94,22 @@ pub struct PackageClaimRoyaltiesInput {}
...
-pub struct PackageDefinition {
-    pub schema: PackageSchema,
-    pub function_access_rules: BTreeMap<String, BTreeMap<String, AccessRule>>,
-    pub royalty_config: BTreeMap<String, RoyaltyConfig>,
+pub struct PackageSetup {
+    pub blueprints: BTreeMap<String, BlueprintSetup>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Default, ScryptoSbor, ManifestSbor)]
+pub struct BlueprintSetup {
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/invocations.rs:102]: `BlueprintDefinition`

File: radix-engine-interface/src/blueprints/package/substates.rs [hunk 4]
```diff
@@ -58,19 +59,13 @@ pub struct PackageRoyaltySubstate {
...
-        IndexedPackageSchema {
-            blueprints: value
-                .blueprints
-                .into_iter()
-                .map(|(name, b)| (name, b.into()))
-                .collect(),
-        }
-    }
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor, ManifestSbor)]
+pub struct BlueprintDefinition {
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/substates.rs:66]: `IndexedBlueprintSchema`

File: radix-engine/src/blueprints/pool/multi_resource_pool/blueprint.rs [hunk 5]
```diff
@@ -232,7 +238,10 @@ impl MultiResourcePoolBlueprint {
             resource_bucket_amount_mapping
         };
 
-        let pool_unit_total_supply = substate.pool_unit_resource_manager.total_supply(api)?;
+        let pool_unit_total_supply = substate
+            .pool_unit_resource_manager
+            .total_supply(api)?
+            .unwrap();
```
Comment (iamyulong) [inline radix-engine/src/blueprints/pool/multi_resource_pool/blueprint.rs:244]: Suggest replacing all the  `.total_supply().unwrap()` with `.expect("Total supply is always enabled for pool unit resource")`

File: radix-engine/src/system/system.rs [hunk 6]
```diff
@@ -306,14 +317,14 @@ where
             self.api
                 .kernel_get_system_state()
                 .system
-                .blueprint_schema_cache
+                .blueprint_cache
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:320]: Minor: slightly prefer `schema/code_cache` based on understanding from our discussion today.

**Review by iamyulong — APPROVED** [review]
Looks good to me. 

Most of the comments are about naming to align with our architecture discussion.

File: radix-engine-interface/src/blueprints/package/invocations.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/invocations.rs:19]: Will keep this for now as I have a separate PR which updates all naming.

File: radix-engine-interface/src/blueprints/package/invocations.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/invocations.rs:108]: Will keep this for now as I have a separate PR which updates all naming.

File: radix-engine-interface/src/blueprints/package/invocations.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/invocations.rs:102]: Will keep this for now as I have a separate PR which updates all naming.

File: radix-engine/src/system/system.rs (hunk 6 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:320]: Will keep this for now as I have a separate PR which updates all naming.

**Pull request merged**